### PR TITLE
[QoI] Add fix-it for missing arguments

### PR DESCRIPTION
**Pull request by rintaro**
Add insertion fix-it with placeholder for missing arguments.

given:
```swift
func fn1(foo : Int, bar: String, baz: Float) {}
fn1(foo: 1, baz: 0.2)

func fn2(foo : Int, bar: (Int) -> Void) {}
fn2 { print($0) }
```

<pre>
<b>test.swift:2:4: error: missing argument for parameter 'bar' in call</b>
fn1(foo: 1, baz: 0.2)
          ^
          , bar: <#String#>
<b>test.swift:5:5: error: missing argument for parameter 'foo' in call</b>
fn2 { print($0) }
   ^
   (foo: <#Int#>)
</pre>

### CODE REVIEW

**Comment by rintaro** [comment]
I'm not sure we want to use *typed* editor placeholder in fix-its.
Anyone have any opinion?

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -4730,13 +4733,115 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
...
+      // Explode inout type.
+      if (auto IOT = param.Ty->getAs<InOutType>()) {
+        insertText << "&";
+        Ty = IOT->getObjectType();
+      }
+      // @autoclosure; the type should be the result type.
+      if (auto FT = param.Ty->getAs<AnyFunctionType>())
+        if (FT->isAutoClosure())
+          Ty = FT->getResult();
+      insertText << "<#T##" << Ty << "#>";
```
Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: IIRC, the syntax for a typed completion token is `<#name##type#>`; there's nothing special about "T". We should then show the internal name of the parameter, like code completion does. @benlangmuir, thoughts?

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -4730,13 +4733,115 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
...
+          assert(PE->hasTrailingClosure() &&
+                 "paren less ParenExpr without trailing closure");
+          insertBuf.insert(insertBuf.begin(), '(');
+          insertBuf.insert(insertBuf.end(), ')');
+          insertLoc = Lexer::getLocForEndOfToken(
+              TC.Context.SourceMgr, FnExpr->getEndLoc());
+        }
+      } else {
+        // Unreachable. Can't be TupleShuffleExpr because this argExpr is not
+        // yet resolved.
```
Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: Can you put either an llvm_unreachable or a return here, then?

**Review by jrose-apple — COMMENTED** [review]
Seems like a good approach. I think putting the fix-it on the error itself is fine, since it is the only valid fix for the issue (other than calling a different method).

On putting typed completion tokens in diagnostics: I've been meaning to poke at the text diagnostic printer to identify these completions and print them in a prettier way…somehow. However, until then it really is a worse command-line experience, so maybe we should go with untyped completion tokens for now. We'll want to do a general pass later anyway if we ever get that feature.

File: lib/Sema/CSDiag.cpp (hunk 1 above)
Comment (rintaro) [inline lib/Sema/CSDiag.cpp]: Here is the syntax of editor placeholder.
https://github.com/apple/swift/blob/master/lib/Basic/EditorPlaceholder.cpp#L24-L34

Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: Oh, whoops, my bad.

**Comment by rintaro** [comment]
I agree about command-line experience.
Updated the PR to use Just `<#TypeName#>`

**Review by jrose-apple — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test and merge

**Comment by rintaro** [comment]
@swift-ci Please smoke test and merge

**Pull request merged**